### PR TITLE
Fix Format parsing in Printf (escaped %)

### LINKED ISSUE #37784: Incorrect escaping in Printf on master (@sprintf("GAP[%%]"))
On master, the Printf library is redisegned but the escape sequences seems not to be hanled correctly (or like in C as written in docs). Now, it is not possible to print single '%' in the middle of string.
``` julia
julia> using Printf
julia> @sprintf("GAP[%%]")
ERROR: LoadError: ArgumentError: invalid format string: 'GAP[%%]', invalid type specifier: ']'
Stacktrace:
 [1] Printf.Format(f::String)
   @ Printf ~/software/julia/julia-dev/usr/share/julia/stdlib/v1.6/Printf/src/Printf.jl:157
 [2] @sprintf(__source__::LineNumberNode, __module__::Module, fmt::Any, args::Vararg{Any, N} where N)
   @ Printf ~/software/julia/julia-dev/usr/share/julia/stdlib/v1.6/Printf/src/Printf.jl:783
in expression starting at REPL[3]:1
```
C style
``` julia
julia> @ccall printf("GAP[%%]\n"::Cstring)::Cvoid
GAP[%]
```

The problem of the new design comes from the commit [#32859](https://github.com/JuliaLang/julia/pull/32859) where the structure in `stdlib/Printf/src/Printf.jl`
``` julia
struct Format{S, T}
    str::S # original full format string as CodeUnits
    # keep track of non-format specifier strings to print
    # length(substringranges) == length(formats) + 1
    # so when printing, we start with printing
      # str[substringranges[1]], then formats[1] + args[1]
      # then str[substringranges[2]], then formats[2] + args[2]
      # and so on, then at the end, str[substringranges[end]]
    substringranges::Vector{UnitRange{Int}}
    formats::T # Tuple of Specs
end
```
expects that `length(substringranges) == length(formats) + 1` but it is not true in general if `"%%"` occurs in the middle of the string format. To provide simple example, there is no format but several substrings in the following:

``` julia
@assert @sprintf("a%%f%%d%%x") == "a%f%d%x"
```
Utilized version of master branch
``` julia
julia> versioninfo()
Julia Version 1.6.0-DEV.unknown
Commit b89c4a3 (2020-09-28 08:12 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i5-1035G1 CPU @ 1.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-10.0.1 (ORCJIT, icelake-client)
```
Btw, all it works well on 1.5.2.

### PR DESCRIPTION
**Pull request by petvana**
This PR aims to fix #37784 by updating `Format(f::AbstractString)` function. It moves format detection into a separate inlined function (only refactoring, no changes) for better readability. There are also new tests for the issue. Notice performance has not been properly tested yet.

### CODE REVIEW

**Comment by petvana** [comment]
Sorry, no idea what caused the error on 32-bit during bootstrap on the new `Printf._detectFormat!` function. 
```
...
Markdown  ���������������������������������  2.073647 seconds
error during bootstrap:
LoadError("sysimg.jl", 19, LoadError("/buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/LibGit2/src/LibGit2.jl", 3, LoadError("/buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/LibGit2/src/signature.jl", 58, LoadError("/buildworker/worker/package_linux32/build/usr/share/julia/stdlib/v1.6/LibGit2/src/signature.jl", 62, MethodError(Printf._detectFormat!, ("%+03i:%02i", UInt8[0x25, 0x2b, 0x30, 0x33, 0x69, 0x3a, 0x25, 0x30, 0x32, 0x69], 10, 2, Printf.Spec[]), 0x000042a9)))))
jl_method_error_bare at /buildworker/worker/package_linux32/build/src/gf.c:1738
jl_method_error at /buildworker/worker/package_linux32/build/src/gf.c:1756
jl_lookup_generic_ at /buildworker/worker/package_linux32/build/src/gf.c:2326 [inlined]
...
```

File: stdlib/Printf/src/Printf.jl [hunk 1]
```diff
@@ -73,115 +73,116 @@ end
...
 char(::Type{Val{c}}) where {c} = c
 
-# parse format string
-function Format(f::AbstractString)
-    isempty(f) && throw(ArgumentError("empty format string"))
-    bytes = codeunits(f)
-    len = length(bytes)
-    pos = 1
-    b = 0x00
+@inline function _detectFormat!(f, bytes, len, pos::Int64, fmts::Vector{Spec})
```
Comment (simeonschaub) [inline stdlib/Printf/src/Printf.jl:76]:
```suggestion
@inline function _detectFormat!(f, bytes, len, pos::Int, fmts::Vector{Spec})
```
That's probably why it fails on 32 bit

**Comment by quinnj** [comment]
Could you walk-through the changes here? It's a bit hard to tell what changed with the refactoring. What was the core issue? What is the fix?

**Comment by petvana** [comment]
> Could you walk-through the changes here? It's a bit hard to tell what changed with the refactoring. What was the core issue? What is the fix?

It's hard to follow mainly because diff at GitHub failed to detect that I have moved ~80% of code unchanged to a separate function `_detectFormat!` with a different indention. The idea behind the code is as follows. The problem was when `"%%]"` occurred. The former code considered `]` as an unsupported format and threw an error.  It would have been possible to fix the code without refactoring into a separate function, but it would lead to messy and hard-to-maintain code.

Now, the proposed code in `Format(f::AbstractString)` parses the format until the end is reached (and the process is terminated by `break`) while each unescaped `%` is processed by `_detectFormat!` function.

Variables `start` and `pos` stands for the current segment that is being parsed, and the variable 'escaped' tells if there is an odd number of `%` in a row - then the format should be detected.

**Comment by StefanKarpinski** [comment]
> t's hard to follow mainly because diff at GitHub failed to detect that I have moved ~80% of code unchanged to a separate function `_detectFormat!` with a different indention.

I would suggest viewing with the "ignore whitespace" box checked: https://github.com/JuliaLang/julia/pull/37807/files?diff=unified&w=1

**Comment by petvana** [comment]
I have tried to create a micro-benchmark focused mainly on format parsing. 

<table class="data-frame"><thead><tr><th></th><th>PrintfVersion</th><th>T1</th><th>T2</th><th>T3</th><th>T4</th><th>T5</th><th>T6</th></tr><tr><th></th><th>String</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th></tr></thead><tbody>
<p>Computational time in nanoseconds measured by `@belapsed`, `NaN` -> failed</p>
<tr><th>1</th><td>JuliaLang/julia/v1.5.2</td><td>51.2989</td><td>240.779</td><td>123.127</td><td>134.039</td><td>469.168</td><td>54.7898</td></tr><tr><th>2</th><td>JuliaLang/julia/master</td><td>22.6024</td><td>307.746</td><td>48.4565</td><td>343.642</td><td>580.486</td><td>NaN</td></tr><tr><th>3</th><td>JuliaLang/julia/jq/37784</td><td>26.7329</td><td>418.95</td><td>48.2624</td><td>351.687</td><td>580.322</td><td>NaN</td></tr><tr><th>4</th><td>petvana/julia/jq/37784</td><td>23.5562</td><td>418.623</td><td>48.2551</td><td>353.719</td><td>597.915</td><td>47.3559</td></tr><tr><th>5</th><td>petvana/julia/Printf-no-unroll</td><td>23.6084</td><td>418.216</td><td>48.3725</td><td>343.617</td><td>539.508</td><td>43.0071</td></tr></tbody></table>

The test instances were as follows
``` julia
# T1
Printf.@sprintf("short");
# T2
Printf.@sprintf("longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong");
# T3 
Printf.@sprintf(" %15d ", 10^6);
# T4
Printf.@sprintf(" %d %d %d %d %d ", 1, 2, 3, 4, 5);
# T5
Printf.@sprintf(" %f %f %f %f %f ", pi, pi, pi, pi, pi);
# T6
Printf.@sprintf(" %% %% %% %% %% %% %% ");
```
Surprisingly cases T2, T4, T5 are slower than in v1.5.2 (at least on my CPU). During playing around, I also found that unrolling in output function `format(buf::Vector{UInt8}, pos::Integer, f::Format, args...)` seems to have a negative influence on the performance, the unrolled version is in branch [petvana/julia/Printf-no-unroll](https://github.com/petvana/julia/tree/Printf-no-unroll) ([DIFF](https://github.com/petvana/julia/compare/jq/37784...petvana:Printf-no-unroll)). @quinnj Do you have any example in which the unrolling is superior?

All the branches have been tested on the same Julia binary: 
```
Julia Version 1.6.0-DEV.1087
Commit 8a9666ae22 (2020-09-29 22:33 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: Intel(R) Core(TM) i5-1035G1 CPU @ 1.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-10.0.1 (ORCJIT, icelake-client)
```

<details>
 <summary>Full Julia source-code to replicate results</summary>

``` julia
module BenchmarkPrintf

using BenchmarkTools
using DataFrames
using Test

# where the tested codes are stored
dir = "versions"

branches = [
    "JuliaLang/julia/v1.5.2",
    "JuliaLang/julia/master",
    "JuliaLang/julia/jq/37784",
    "petvana/julia/jq/37784",
    "petvana/julia/Printf-no-unroll",
]

#rm(dir, recursive=true, force=true)

data = DataFrame()
data.PrintfVersion = branches
data.T1 = NaN
data.T2 = NaN
data.T3 = NaN
data.T4 = NaN
data.T5 = NaN
data.T6 = NaN

for (idx, name) in enumerate(branches)
    @info "Benchmarking $(name)"
    actdir = dir * "/" * replace(name, "/" => "-")
    source_file = actdir * "/Printf.jl"
    mkpath(actdir)

    if name == "local"
        # cp(local_file, source_file, force=true)
    else
        url = "https://raw.githubusercontent.com/$(name)/stdlib/Printf/src/Printf.jl"
        if !isfile(source_file)
            @info "Downloading from $(url)"
            download(url, source_file)
            run(`sed -i 's/using Base.Grisu/using Grisu/g' $(source_file)`)
        end
    end

    include(source_file)
    using .Printf

    instance = :T1
    try
        t = @belapsed  Printf.@sprintf("short");
        data[idx, instance] = t * 10^9
    catch
    end

    instance = :T2
    try
        t = @belapsed Printf.@sprintf("longlonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglonglong");
        data[idx, instance] = t * 10^9
    catch
    end

    instance = :T3
    try
        t = @belapsed  Printf.@sprintf(" %15d ", 10^6);
        data[idx, instance] = t * 10^9
    catch
    end

    instance = :T4
    try
        t = @belapsed  Printf.@sprintf(" %d %d %d %d %d ", 1, 2, 3, 4, 5);
        data[idx, instance] = t * 10^9
    catch
    end

    instance = :T5
    try
        t = @belapsed  Printf.@sprintf(" %f %f %f %f %f ", pi, pi, pi, pi, pi);
        data[idx, instance] = t * 10^9
    catch
    end   

    instance = :T6
    try
        t = @belapsed  Printf.@sprintf(" %% %% %% %% %% %% %% ");
        data[idx, instance] = t * 10^9
    catch
    end 

    println(data)
end

println("")
show(stdout, MIME("text/html"), data)
println("")
println("")

end
```
</details>

**Comment by petvana** [comment]
To take it more seriously, I have prepared a benchmark closer to real usage. It prints **1 or 10^7 lines into `\dev\null`** containing random float/decimal values. The results indicate that the proposed PR provides at least the same performance as the master branch. I have also removed unrolling because it seems not to give any performance gain, and it is much slower to print the first line (**up to 3.4s for F4** that is 10-times slower). I considered to open a separate PR for removing the unroll, but I didn't want to split the discussion into two threads. Source codes to reproduce the results are in [petvana/BenchmarkPrintf](https://github.com/petvana/BenchmarkPrintf).

@quinnj What do you think about the PR and results?

Utilized formats:
``` julia
# F1
" %f \n",
# F2
" %f %f \n"
# F3
" %f %f %f %f \n",
# F4
" %f %f %f %f %f %f %f %f \n"
# F5
" Influence %f [%%], data %f \n"
# F6
" Influence %f [%%], data %f, %f, %f, %f, %f, %f, %f, \n"
# F7
" Very very very very very very very very very very very very very very very very very very very very long text %f \n"

# D1
" %d \n",
# D2
" %d %d \n"
# D3
" %d %d %d %d \n",
# D4
" %d %d %d %d %d %d %d %d \n"
# D5
" Influence %d [%%], data %d \n"
# D6
" Influence %d [%%], data %d, %d, %d, %d, %d, %d, %d, \n"
# D7
" Very very very very very very very very very very very very very very very very very very very very long text %d \n"
```

<table class="data-frame"><thead><tr><th></th><th>PrintfVersion</th><th>F1</th><th>F2</th><th>F3</th><th>F4</th><th>F5</th><th>F6</th><th>F7</th></tr><tr><th></th><th>String</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th></tr></thead><tbody><p>Print 1 line with float value(s) [s]</p><tr><th>1</th><td>JuliaLang/julia/v1.5.2</td><td>0.898296</td><td>0.0825418</td><td>0.0801878</td><td>0.140242</td><td>0.0476642</td><td>0.137446</td><td>0.0301876</td></tr><tr><th>2</th><td>JuliaLang/julia/master</td><td>0.366507</td><td>0.531089</td><td>1.3461</td><td><b>3.41591</b></td><td>0.0173058</td><td>0.0233405</td><td>0.0144985</td></tr><tr><th>3</th><td>JuliaLang/julia/jq/37784</td><td>0.311374</td><td>0.533566</td><td>1.26168</td><td><b>3.11948</b></td><td>0.0174601</td><td>0.0201481</td><td>0.0154846</td></tr><tr><th>4</th><td>petvana/julia/jq/37784</td><td>0.308301</td><td>0.317726</td><td>0.318067</td><td>0.33422</td><td>0.0163931</td><td>0.0201942</td><td>0.014353</td></tr></tbody></table>

<table class="data-frame"><thead><tr><th></th><th>PrintfVersion</th><th>F1</th><th>F2</th><th>F3</th><th>F4</th><th>F5</th><th>F6</th><th>F7</th></tr><tr><th></th><th>String</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th></tr></thead><tbody><p>Print 10^7 lines with float value(s) [s]</p><tr><th>1</th><td>JuliaLang/julia/v1.5.2</td><td>5.74427</td><td>10.9207</td><td>19.5306</td><td>33.2879</td><td>10.8054</td><td>33.8352</td><td>4.74343</td></tr><tr><th>2</th><td>JuliaLang/julia/master</td><td>3.7729</td><td>7.68967</td><td>17.5296</td><td>25.8437</td><td>7.29169</td><td>23.3581</td><td>4.76526</td></tr><tr><th>3</th><td>JuliaLang/julia/jq/37784</td><td>3.75447</td><td>7.66193</td><td>17.3082</td><td>25.8387</td><td>7.46661</td><td>23.0917</td><td>5.39984</td></tr><tr><th>4</th><td>petvana/julia/jq/37784</td><td>4.09708</td><td>7.50588</td><td>16.7598</td><td>22.0186</td><td>7.78205</td><td>22.0153</td><td>5.62088</td></tr></tbody></table>

<table class="data-frame"><thead><tr><th></th><th>PrintfVersion</th><th>D1</th><th>D2</th><th>D3</th><th>D4</th><th>D5</th><th>D6</th><th>D7</th></tr><tr><th></th><th>String</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th></tr></thead><tbody><p>Print 1 line with integer value(s) [s]</p><tr><th>1</th><td>JuliaLang/julia/v1.5.2</td><td>0.0441159</td><td>0.0348119</td><td>0.0508356</td><td>0.0818782</td><td>0.0345189</td><td>0.0796283</td><td>0.0235023</td></tr><tr><th>2</th><td>JuliaLang/julia/master</td><td>0.131236</td><td>0.108799</td><td>0.195427</td><td>0.365146</td><td>0.0177598</td><td>0.0213297</td><td>0.0159883</td></tr><tr><th>3</th><td>JuliaLang/julia/jq/37784</td><td>0.146931</td><td>0.123761</td><td>0.223365</td><td>0.399527</td><td>0.0193841</td><td>0.0232306</td><td>0.0163871</td></tr><tr><th>4</th><td>petvana/julia/jq/37784</td><td>0.147011</td><td>0.0921646</td><td>0.11106</td><td>0.119831</td><td>0.0181344</td><td>0.0219818</td><td>0.0167515</td></tr></tbody></table>

<table class="data-frame"><thead><tr><th></th><th>PrintfVersion</th><th>D1</th><th>D2</th><th>D3</th><th>D4</th><th>D5</th><th>D6</th><th>D7</th></tr><tr><th></th><th>String</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th><th>Float64</th></tr></thead><tbody><p>Print 10^7 lines with integer value(s) [s]</p><tr><th>1</th><td>JuliaLang/julia/v1.5.2</td><td>6.8859</td><td>16.0583</td><td>28.9534</td><td>54.3109</td><td>15.5503</td><td>53.4028</td><td>7.74338</td></tr><tr><th>2</th><td>JuliaLang/julia/master</td><td>2.55479</td><td>4.60268</td><td>10.9064</td><td>17.2639</td><td>5.22935</td><td>17.1426</td><td>3.69654</td></tr><tr><th>3</th><td>JuliaLang/julia/jq/37784</td><td>2.6154</td><td>4.87605</td><td>10.9827</td><td>17.555</td><td>5.31121</td><td>17.38</td><td>4.3217</td></tr><tr><th>4</th><td>petvana/julia/jq/37784</td><td>3.04381</td><td>4.9267</td><td>11.0332</td><td>16.0263</td><td>5.01531</td><td>16.5788</td><td>4.00052</td></tr></tbody></table>

**Comment by quinnj** [comment]
Going to try and dig into this now; sorry for the delay. Those benchmarks are.......a little weird since they seem like very obscure/corner case uses of printf. Most of the benchmarking I've done is just regular floats/ints/strings and sometimes mixed. The unrolling code should be much faster in the mixed type args case. The only benchmark you posted that really concerns me is T4 where printing 5 integers seems to be slower than 1.5.2.

**Comment by quinnj** [comment]
Ok, I pushed another simple commit that fixes the rest of the escaping test cases (and adds those tests) (https://github.com/JuliaLang/julia/pull/37785/commits/d2f65dbb748f908a28475d24936c244c6bb4b3e1).

I do appreciate the performance benchmarking you've done across all these cases @petvana; I think that's a bit outside of the scope of the specific fix for the original issue (the fix doesn't itself introduce any performance issues). So let's merge the fixes in my branch and maybe open a new issue for the problematic performance issues you've found.

**Comment by petvana** [comment]
Thank you for the fix @quinnj. I have gone through the code, and it seems to solve all possible cases. 

I have also tested performance with mixed types, and I can confirm the unrolling code is much faster in such a scenario. It seems the compiler can optimize (unroll) the code automatically if all the types are the same.

Now, the only performance drop (10-13%) of your branch compare to master is for very long substring ranges (F7, D7), but I guess they are sporadic. An alternative would be to pre-process the format string and remove escaped '%' symbols during Format parsing (compile time), but it would be necessary to store the modified format as a copy.

The reason for the performance drop in T4 (compare to 1.5.2) is still unknown to me, but it can be somehow related to the fact that the inputs are constant.

I will close this PR for now, as the bug is fixed.

**Comment by quinnj** [comment]
Thanks for the response @petvana; I looked a little into the performance of T4 last night and I _think_ it has to do with our creating a `StringVector` on each call of `@sprintf` (profiling shows most of the time is spent allocating the `StringVector`, then `resize!`ing at the end). The benchmarks are much better when comparing the `Printf.format(buf, pos, fmt, args...)` method, and reusing the same `buf` each time. That's good because, for example, in `CSV.write`, we would manage our own `buf` and use this method to avoid the extra allocations. So I'm not sure for the single `@sprintf` case how big of a deal it really is. The old grisu/printf code I think had a bunch of global buffers that it reused (which led to at least a couple of bugs that I remember), so there's also a tradeoff there: we could reuse global buffers and probably get some better performance, but it's a non-trivial amount of work and could introduce subtle bugs.
